x-pack/elastic-agent/pkg/composable/providers/kubernetes/node.go

```diff
@@ -0,0 +1,209 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package kubernetes
+
+import (
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	k8s "k8s.io/client-go/kubernetes"
+
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/libbeat/common/kubernetes"
+	"github.com/elastic/beats/v7/libbeat/common/safemapstr"
+	"github.com/elastic/beats/v7/libbeat/logp"
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/agent/errors"
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/composable"
+)
+
+type node struct {
+	logger         *logp.Logger
+	cleanupTimeout time.Duration
+	comm           composable.DynamicProviderComm
+	scope          string
+}
+
+type nodeData struct {
+	node       *kubernetes.Node
+	mapping    map[string]interface{}
+	processors []map[string]interface{}
+}
+
+// NewNodeWatcher creates a watcher that can discover and process node objects
+func NewNodeWatcher(
+	comm composable.DynamicProviderComm,
+	cfg *ResourceConfig,
+	logger *logp.Logger,
+	client k8s.Interface,
+	scope string) (kubernetes.Watcher, error) {
+	watcher, err := kubernetes.NewWatcher(client, &kubernetes.Node{}, kubernetes.WatchOptions{
+		SyncTimeout:  cfg.SyncPeriod,
+		Node:         cfg.Node,
+		IsUpdated:    isUpdated,
+		HonorReSyncs: true,
+	}, nil)
+	if err != nil {
+		return nil, errors.New(err, "couldn't create kubernetes watcher")
+	}
+	watcher.AddEventHandler(&node{logger, cfg.CleanupTimeout, comm, scope})
+
+	return watcher, nil
+}
+
+func (n *node) emitRunning(node *kubernetes.Node) {
+	data := generateNodeData(node)
+	data.mapping["scope"] = n.scope
+	if data == nil {
```

Comment:
Can this happen taking the previous line into account?

Reply:
Good catch